File: 2021/src/day05/day05.js
const line = (start, end) => {
  const horizontal = start[0] === end[0];
  const vertical = start[1] === end[1];

  const deltaH = end[0] - start[0];
  const deltaV = end[1] - start[1];

  const slope = [];
  if (vertical) {
    slope[0] = deltaH / Math.abs(deltaH);
    slope[1] = 0;
  } else if (horizontal) {
    slope[0] = 0;
    slope[1] = deltaV / Math.abs(deltaV);
  } else {
    slope[0] = deltaH > 0 ? 1 : -1;
    slope[1] = deltaV > 0 ? 1 : -1;
  }

  const allPoints = () => {
    const points = [start];
    let x = start[0];
    let y = start[1];
    while (x !== end[0] || y !== end[1]) {
      x += slope[0];
      y += slope[1];
      points.push([x, y]);
    }
    return points;
  };

  return { allPoints, end, horizontal, start, vertical };
};

const parseInput = (raw) =>
  raw
    .trim()
    .split("\n")
    .map((v) => {
      const [start, end] = v.split(" -> ");
      return line(start.split(",").map(Number), end.split(",").map(Number));
    });

const findIntersections = (input) => {
  const map = new Map();
  input.forEach((line) => {
    line.allPoints().forEach(([x, y]) => {
      const key = `${x},${y}`;
      if (!map.has(key)) {
        map.set(key, 0);
      }
      map.set(key, map.get(key) + 1);
    });
  });
  return Array.from(map.values()).filter((v) => v > 1).length;
};

export const part1 = (raw) => {
  const input = parseInput(raw).filter(
    ({ horizontal, vertical }) => horizontal || vertical,
  );

  return findIntersections(input);
};

export const part2 = (raw) => {
  const input = parseInput(raw);

  return findIntersections(input);
};

File: 2021/src/day05/day05.swift
import Foundation

private struct Point: Equatable, Hashable {
  var x: Int
  var y: Int

  static func == (lhs: Point, rhs: Point) -> Bool {
    return lhs.x == rhs.x && lhs.y == rhs.y
  }

  func hash(into hasher: inout Hasher) {
    hasher.combine(x)
    hasher.combine(y)
  }
}

private class Line {
  private var start: Point
  private var end: Point

  init(with lineStr: String) {
    let coords = lineStr.replacingOccurrences(of: " -> ", with: ",").split(separator: ",")
    start = Point(x: Int(coords[0])!, y: Int(coords[1])!)
    end = Point(x: Int(coords[2])!, y: Int(coords[3])!)
  }

  func allPoints() -> [Point] {
    if horizontal() {
      let range = start.x < end.x ? start.x ... end.x : end.x ... start.x
      return range.map { Point(x: $0, y: start.y) }
    } else if vertical() {
      let range = start.y < end.y ? start.y ... end.y : end.y ... start.y
      return range.map { Point(x: start.x, y: $0) }
    } else {
      var points: [Point] = []

      var deltaX = 0
      var deltaY = 0

      deltaX = end.x - start.x > 0 ? 1 : -1
      deltaY = end.y - start.y > 0 ? 1 : -1

      var x = start.x
      var y = start.y

      while x != end.x || y != end.y {
        points.append(Point(x: x, y: y))
        x += deltaX
        y += deltaY
      }
      points.append(Point(x: x, y: y))

      return points
    }
  }

  func horizontal() -> Bool {
    return start.y == end.y
  }

  func vertical() -> Bool {
    return start.x == end.x
  }
}

enum Day5 {
  private static func parse(_ contents: String) -> [Line] {
    return contents.split(separator: "\n").map { Line(with: String($0)) }
  }

  private static func run(ventLines: [Line]) -> Int {
    var board: [Point: Int] = [:]

    for vent in ventLines {
      for point in vent.allPoints() {
        if let count = board[point] {
          board[point] = count + 1
        } else {
          board[point] = 1
        }
      }
    }
    return board.values.filter { $0 > 1 }.count
  }

  static func part1(input: String) -> Int {
    let data = parse(input).filter { $0.horizontal() || $0.vertical() }
    return run(ventLines: data)
  }

  static func part2(input: String) -> Int {
    let data = parse(input)
    return run(ventLines: data)
  }
}
